contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,384 @@
+            uint256 uniAmount = tokenAvailable(uniFactory, address(_inputToken)) ? getSellPrice(true, address(_inputToken), _amountInput) : 0;
+            uint256 sushiAmount = tokenAvailable(sushiFactory, address(_inputToken)) ? getSellPrice(false, address(_inputToken), _amountInput) : 0; 
+            amountEth = Math.max(uniAmount, sushiAmount);
+        } else {
+            amountEth = _amountInput;
+        }
+
+        // get price of set token on uniswap
+        uint256 wethBalance = amountEth;
+        (uint256 tokenReserveA, uint256 tokenReserveB) = UniswapV2Library.getReserves(uniFactory, WETH, address(_setToken));
+        uint256 minSetTokenAmountOut = uniRouter.getAmountOut(wethBalance, tokenReserveA, tokenReserveB);
+
+        uint256 sumEth = 0;
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        uint256[] memory amountEthIn = new uint256[](positions.length);
+        for(uint256 i = 0; i < positions.length; i++) {
+            uint256 unit = uint256(positions[i].unit);       // amount of token per 1 index token
+            address token = positions[i].component;
+            uint256 amountOut = minSetTokenAmountOut * unit / 1 ether;
+            uint256 uniPrice = tokenAvailable(uniFactory, token) ? getBuyPrice(true, token, amountOut) : PreciseUnitMath.maxUint256();
+            uint256 sushiPrice = tokenAvailable(sushiFactory, token) ? getBuyPrice(false, token, amountOut) : PreciseUnitMath.maxUint256();
+            uint256 amountEth = Math.min(uniPrice, sushiPrice);
+            sumEth += amountEth;
+            amountEthIn[i] = amountEth;
+        }
+
+        uint256 maxIndexAmount = PreciseUnitMath.maxUint256();
+        for (uint i = 0; i < positions.length; i++) {
+            address token = positions[i].component;
+            uint256 unit = uint256(positions[i].unit);
+
+            uint256 scaledAmountEth = (amountEthIn[i] * wethBalance) / sumEth;  // scale the amountEthIn
+            uint256 uniTokenOut = tokenAvailable(uniFactory, token) ? getBuyPriceExactETH(true, token, scaledAmountEth) : 0;
+            uint256 sushiTokenOut = tokenAvailable(sushiFactory, token) ? getBuyPriceExactETH(false, token, scaledAmountEth) : 0;
+
+            uint256 amountTokenOut = Math.max(uniTokenOut, sushiTokenOut);
+            // update the maxIndexAmount
+            maxIndexAmount = Math.min(amountTokenOut * 1 ether / unit, maxIndexAmount);
+        }
+        return maxIndexAmount;
+    }
+
+    /* ============ Internal Functions ============ */
+
+    function liquidateComponents(ISetToken _setToken) internal {
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        for (uint256 i=0; i<positions.length; i++) {
+            sellTokenBestPrice(positions[i].component);
+        }
+    }
+
+    function acquireComponents(ISetToken.Position[] memory positions, uint256[] memory amountEthIn, uint256 wethBalance, uint256 sumEth) internal returns (uint256) {
```

Comment:
In conformity with the rest of this file, please prefix parameters with underscore!

If an internal function has a return var, it'd be great to have a javadoc to describe it.